openupgradelib/openupgrade_160.py

```diff
@@ -45,3 +58,346 @@ def migrate_translations_to_jsonb(env, fields_spec):
                 if initial_translation_tables == "ir_translation":
                     query = query.replace("_ir_translation", "ir_translation")
                 logged_query(env.cr, query)
+
+
+_BADGE_CONTEXTS = (
+    "secondary",
+    "primary",
+    "success",
+    "info",
+    "warning",
+    "danger",
+)
+
+_RTL_REPLACEMENT_CONTEXT = (
+    ("left", "start"),
+    ("right", "end"),
+)
+
+_RTL_REPLACEMENT_ELEMENT = (
+    "text",
+    "float",
+    "border",
+    "border-top",
+    "border-bottom",
+    "rounded",
+)
+
+_MARGIN_PADDING_ELEMENT_REPLACEMENT = (
+    ("pl", "ps"),
+    ("ml", "ms"),
+    ("pr", "pr"),
+    ("mr", "me"),
+)
+
+_MARGIN_PADDING_SIZE = ("0", "1", "2", "3", "4", "5", "auto")
+
+_MARGIN_PADDING = ("sm", "lg")
+
+# These replacements are from standard Bootstrap 4 to 5
+_BS5_REPLACEMENTS = (
+    # Grid stuff
+    _r(class_rm="no-gutters", class_add="g-0"),
+    _r(class_rm="media", class_add="d-flex"),
+    _r("media-body", "flex-grow-1"),
+    # Content, Reboot, etc
+    _r("thead-light", "table-light"),
+    _r("thead-dark", "table-dark"),
+    # Special case where text-justify no longer exist
+    _class_rp_by_inline(
+        selector="//*[contains(@class, 'text-justify')]",
+        selector_mode="xpath",
+        class_rp_by_inline={"text-justify": ["text-align: justify"]},
+    ),
+    _r(class_rm="text-justify", class_add=""),
+    # RTL
+    *(
+        _r("%s-%s" % (elem, t4), "%s-%s" % (elem, t5))
+        for (t4, t5), elem in product(
+            _RTL_REPLACEMENT_CONTEXT, _RTL_REPLACEMENT_ELEMENT
+        )
+    ),
+    _r("pl", "ps"),
+    _r("pr", "pe"),
+    _r("ml", "ms"),
+    _r("mr", "me"),
+    # For stub like pl-0 -> ps-0
+    *(
+        _r("%s-%s" % (t4, size), "%s-%s" % (t5, size))
+        for (t4, t5), size in product(
+            _MARGIN_PADDING_ELEMENT_REPLACEMENT, _MARGIN_PADDING_SIZE
+        )
+    ),
+    # For stub like ml-sm-1 -> ms-sm-1
+    *(
+        _r("%s-%s-%s" % (t4, context, size), "%s-%s-%s" % (t5, context, size))
+        for (t4, t5), context, size in product(
+            _MARGIN_PADDING_ELEMENT_REPLACEMENT, _MARGIN_PADDING, _MARGIN_PADDING_SIZE
+        )
+    ),
+    # Forms
+    _r("custom-control", "form-control"),
+    _r("custom-checkbox", "form-check"),
```

Comment:
@duong77476 trường hợp `custom-checkbox` thì cần chuyển cả  `custom-checkbox custom-control` thành `form-check`. Để như này nó sẽ thành `form-check form-control` gây lỗi

![Screenshot from 2023-09-07 16-07-00](https://github.com/Viindoo/openupgradelib/assets/66666640/a764de3c-c98b-40bc-b60b-148857fbcbec)

Reply:
ok anh để e xử ở con master của em

Reply:
@ngochuy97hp Kiểu như này đúng ko a
`_r(
        selector=".custom-checkbox + .custom-control",
        class_rm="custom-checkbox",
        class_add="form-check"
    ),
    `
    e chưa test , thấy thằng khác dùng như thế thôi
cc @royleviindoo

Reply:
à có khi ko cần dấu `+`

Reply:
Bắt chước ở https://github.com/OCA/openupgradelib/blob/f5be91a5c77220285af04e9b0d1d3bc4273b33e9/openupgradelib/openupgrade_120.py#L118
cc @royleviindoo @ngochuy97hp

Reply:
anh chưa đọc cái hàm `_r` này miễn là 2 thằng này `custom-checkbox custom-control` biến thành `form-check` là được

Reply:
@ngochuy97hp oh ok a, vậy nó nên là
`_r(
        selector=".custom-checkbox .custom-control",
        class_rm="custom-checkbox custom-control",
        class_add="form-check",
    ),`
    
nghĩa là selector đến element nào có 2 class trên, xóa 2 thk đó thay bằng thk `form-check`

Reply:
Done in https://github.com/duong77476/openupgradelib/pull/7

Reply:
uh đúng r e